Shutdown masters more gracefully

muaazsaleem opened the pull request:
### Wait to serve current requests before terminating api-server

To avoid spiking error rates for clients, take api-server out of
the kubernetes.default endpoint and then wait while the current
requests are being served

ref: https://github.com/kubernetes/kubernetes/pull/74416

Signed-off-by: Muaaz Saleem <muhammad.muaaz.saleem@zalando.de>

aermakov-zalando commented on cluster/node-pools/master-default/userdata.yaml:
```diff
@@ -97,6 +97,10 @@ write_files:
           image: nonexistent.zalan.do/teapot/kube-apiserver:fixed
           args:
           - --apiserver-count={{ .Values.apiserver_count }}
+          {{- if eq .Cluster.ConfigItems.endpoint_reconciliation "false" }}
```
This will still roll the nodes I think because of the extra spaced.

aermakov-zalando commented on cluster/manifests/kube-api-server/kuberntes-service.yaml:
```diff
@@ -0,0 +1,40 @@
+  type: ExternalName
+---
+{{end}}
+apiVersion: v1
```
Split into another file maybe?

aermakov-zalando commented on cluster/manifests/kube-api-server/kuberntes-service.yaml:
```diff
@@ -0,0 +1,40 @@
+    targetPort: 443
+  selector:
+    application: kube-apiserver
+  sessionAffinity: None
```
Should we copy the settings from `kubernetes.default`?

aermakov-zalando commented on cluster/manifests/kube-api-server/kuberntes-service.yaml:
```diff
@@ -0,0 +1,40 @@
+# Since even the "lease" based endpoint reconciler does not shutdown
```
Typo in the filename (`kuberntes`).

aermakov-zalando commented on cluster/config-defaults.yaml:
```diff
@@ -1,3 +1,6 @@
+# temporary feature toggle for the "kubernetes" endpoint reconciler, default is "lease", otherwise "none"
+endpoint_reconciliation: "true"
```
You need two feature toggles (one to rollout the updated service, another one to disable the reconciler). This will disable the reconciler first but rollout the service later.

aermakov-zalando commented on cluster/manifests/kube-api-server/kuberntes-service.yaml:
```diff
@@ -0,0 +1,40 @@
+    port: 443
+    protocol: TCP
+    targetPort: 443
+  sessionAffinity: None
```
Same question here: do we maybe need it?

linki commented on cluster/node-pools/master-default/userdata.yaml:
```diff
@@ -155,6 +159,11 @@ write_files:
+          lifecycle:
+            preStop:
+              exec:
+                command: ["/bin/sh", "-c",  " sleep 60"]
```
There's also the `--shutdown-delay-duration` flag which might help here:

> Time to delay the termination. During that time the server keeps serving requests normally and /healthz returns success, but /readzy immediately returns failure. Graceful termination starts after this delay has elapsed. This can be used to allow load balancer to stop sending traffic to this server.

muaazsaleem replied:
`--shutdown-delay-duration` doesn't help here because it's designed to give applications more time to shutdown. Our problem is that the api-server itself doesn't handle shutdown properly. So we're stopping new connections from coming in and waiting for connections to drain manually.

linki replied:
I think that's what it does: I started apiserver with with `--shutdown-delay-duration=60s` and when I send it a `kill` it immediately takes the IP out of the endpoints list while the apiserver itself keeps running for another minute.

Only available from 1.16.

linki replied:
For reference: https://github.com/kubernetes/kubernetes/pull/74416

aermakov-zalando replied:
Nice! If this works, then we don't even need the rest of it.

muaazsaleem replied:
Just tested `--shutdown-delay-duration` and it works! I think I was confusing it with `terminationGracePeriodSeconds ` in [PodSpec]( https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#pod-v1-core)

Commit:
Wait to serve current requests before terminating api-server

To avoid spiking error rates for clients, take api-server out of
the kubernetes.default endpoint and then wait while the current
requests are being served

Signed-off-by: Muaaz Saleem <muhammad.muaaz.saleem@zalando.de>

muaazsaleem commented:
👍

mikkeloscar commented:
:+1:
